### PR TITLE
[dynamo] Validate check_fn

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130159
* __->__ #118448

Fixes - https://github.com/pytorch/pytorch/issues/128090

Tracker issue here - https://github.com/pytorch/pytorch/issues/129937

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
Test errors?  Are you looking for feedback here?

**Comment by anijain2305** [comment]
I thought that there were only a couple of category of errors and my stack fixed them, but there are just too many :( I will debug and send fixes for others.

**Comment by anijain2305** [comment]
Categorizing the errors. All these require specific knowledge, so asking for help

1) `pytest -vs test/dynamo/test_dynamic_shapes.py::DynamicShapesMiscTests::test_getset_descriptor_dynamic_shapes` - cc @mlazos. This is because the `id(torch.Tensor.shape.__get__)` changes invocation to invocation (idk why). Can you please follow up?

2) `pytest -s test/dynamo/test_higher_order_ops.py::HigherOrderOpVmapGuardTests::test_vmap_guard_fail` - This is vmap + dynamo issue - tracked here https://github.com/pytorch/pytorch/issues/118493

3) `pytest test/dynamo/test_repros.py::ReproTests::test_odict_get_item_index_name` - This happens because of `list(dict.keys()) == [..]` guard. @lezcano  is already working on removing it. Many tests under `quantization` are due to the same issue.

4) `python -s test/dynamo/test_export.py -k test_export_with_symbool_inputs` - This one likely fails because running `check_fn` runs guard `cast_symbool_to_symint_guardless` which upsets the testing code. cc @ydwu4 @ezyang for this one.

5) `python distributions/test_distributions.py::TestJit::test_variance` - Doesnt repro on my machine.

**Review by ezyang — APPROVED** [review]
The assert message when this fails is going to be pretty bad, maybe at minimum print the guard code

**Comment by ezyang** [comment]
Oops i need to read comments lol

**Comment by jansel** [comment]
I'm kind of surprised we didn't find this earlier... won't it just result in a recompile loop until we hit the cache limit?

**Comment by anijain2305** [comment]
> I'm kind of surprised we didn't find this earlier... won't it just result in a recompile loop until we hit the cache limit?

The tests just invoke the `opt_fn` once. So, we never reach cache limit.

**Comment by ezyang** [comment]
Re (4), I don't remember exactly how @ydwu4 implemented this, but one thing I remember discussing in PR review was that guard checks on inflowing SymInts must do checks entirely on hints, and only force guards on the underlying SymInts if the guard actually matches. Assuming that there is a mechanism like this, then to validate check fn, what you want to do is do the hint check but omit the part where guards are actually forced. Not sure how to spell this, but morally that's what should be done. Hopefully @ydwu4 can elaborate.

**Comment by ydwu4** [comment]
I can take a deeper look at the (4). It could be that we should return the hint of the symbolic expression (i.e. a concrete boolean value) instead of the expression itself. I'll sync with @anijain2305 later today.

**Comment by ydwu4** [comment]
For 4), discussed with @anijain2305. The test is expecting dynamo tracing doesn't specialize symbool's shape_env. However, in this pr, check_fn is called immediately, which specializes the outer shape_env of `L['pred']` because of the `if cast_symbool_to_symint(L['pred']) == False `check in the check_fn.

I think it might be the correct behavior to call check_fn immediately to specialize the outer shape_env and we just need to modify the expected guards of outer shape_env. Because the symbool inputs are indeed specialized for this particular code path that dynamo traces and guards. By propagating the guard immediately, it is also consistent with the behavior where we have a symbool inputs for a eager program and we run it without torch.compile, where it specializes the symbool's shape_env. Any comments? @ezyang

**Comment by ezyang** [comment]
Although I agree it all works out in the end, in this case, I would prefer for us to keep the invariant that no updates to the enclosing Python runtime can be made until we are DONE with compiling. Prematurely modifying the Python runtime state during compilation is an invariant violation.

**Comment by anijain2305** [comment]
Another issue discovered at https://github.com/pytorch/pytorch/issues/119338

File: test/dynamo/test_subclasses.py [hunk 1]
```diff
@@ -1363,6 +1363,9 @@ def f(x):
         s = SubTensor(torch.randn(3, 10))
         f(s)
 
+    # Guard validation upsets the guard
+    # https://github.com/pytorch/pytorch/issues/129936
+    @unittest.expectedFailure
```
Comment (anijain2305) [inline test/dynamo/test_subclasses.py:1368]: @ydwu4 Another xfail here. Let me know if thats ok.

File: test/export/test_export.py [hunk 2]
```diff
@@ -4902,6 +4902,9 @@ def forward(self, x, y):
             )
         )
 
+    # Guard validation upsets the guard
+    # https://github.com/pytorch/pytorch/issues/129939
+    @unittest.expectedFailure
```
Comment (anijain2305) [inline test/export/test_export.py:4907]: @ydwu4 xfail here.

File: test/dynamo/test_subclasses.py (hunk 1 above)
Comment (ydwu4) [reply test/dynamo/test_subclasses.py:1368]: Discussed offline. I'll take a look.

**Comment by anijain2305** [comment]
@jansel @ezyang This is ready for review.

Many tests are xfailed. Fixing them would take some time and help from other engineers on the torch.compile stack. They are covered by the tracker.But, I think xfailing and merging this PR is better than waiting.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**